### PR TITLE
Payments/1 items

### PR DESCRIPTION
**Pull request by dunkOnIT**
Still in draft, will add some editorial soon

(also I realize I've completely bastardized payment_denomination, working on restoring it now)

### CODE REVIEW

File: app/models/registration.rb [hunk 1]
```diff
@@ -176,6 +189,14 @@ def to_be_paid_through_wca?
     !new_record? && (pending? || accepted?) && competition.using_payment_integrations? && outstanding_entry_fees > 0
   end
 
+  def invoice_items_total
+    Money.new(invoice_items.sum(:amount_lowest_denomination), entry_fee.currency)
+  end
+
+  def invoice_items_currency_code
```
Comment (dunkOnIT) [inline app/models/registration.rb:196]: This is a little awkward but seems like the best I can do until I can store the currency code in the Order? 

Unless I just call `self.competition.currency_code`

File: db/schema.rb [hunk 2]
```diff
@@ -750,6 +759,21 @@
...
+  end
+
+  create_table "manual_payment_records", charset: "utf8mb4", collation: "utf8mb4_unicode_ci", force: :cascade do |t|
+    t.string "payment_reference", null: false
+    t.bigint "registration_id", null: false
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["registration_id"], name: "index_manual_payment_records_on_registration_id"
+  end
+
```
Comment (gregorbg) [inline db/schema.rb:776]: :upside_down_face:

File: app/models/registration.rb (hunk 1 above)
Comment (gregorbg) [reply app/models/registration.rb:184]: In line with my other comment above, I think just calling the competition's currency code is fine, since that is what effectively defined every invoice item's currency in your current implementation anyways.

**Review by gregorbg — CHANGES_REQUESTED** [review]
I definitely need to see your original draft as a reference to make some more informed decisions about this PR. Can you post the link to your diagram somewhere here or on Slack, please?

File: app/controllers/api/v1/registrations/registrations_controller.rb [hunk 3]
```diff
@@ -230,9 +230,17 @@ def validate_payment_ticket_request
 
   def payment_ticket
     iso_donation_amount = params[:iso_donation_amount].to_i || 0
-    # We could delegate this call to the prepare_intent function given that we're already giving it registration - however,
-    # in the long-term we want to decouple registrations from payments, so I'm deliberately not introducing any more tight coupling
-    ruby_money = @registration.entry_fee_with_donation(iso_donation_amount)
+
+    if Rails.env.production? && EnvConfig.WCA_LIVE_SITE?
```
[the remark this answers is not in the document]
Comment (dunkOnIT) [reply app/controllers/api/v1/registrations/registrations_controller.rb:234]: I'm honestly not sure. I think I added it when I started this PR on the assumption that we wouldn't want to change the prod payment logic just yet - but I'm comfortable with the PR going into prod in its current state. I'll remove it on the assumption that's where this question is leading

File: app/models/registration.rb [hunk 4]
```diff
@@ -176,6 +177,48 @@ def to_be_paid_through_wca?
     !new_record? && (pending? || accepted?) && competition.using_payment_integrations? && outstanding_entry_fees > 0
   end
 
+  def invoice_items_total
+    Money.new(invoice_items.sum { it.amount_lowest_denomination }, entry_fee.currency)
```
[the remark this answers is not in the document]
Comment (dunkOnIT) [reply app/models/registration.rb:181]: Can't remember - hopefully a test will fail and prove that I'm not crazy

File: app/models/registration.rb (hunk 1 above)
Comment (dunkOnIT) [reply app/models/registration.rb:184]: Removed - turns out it wasn't used anyway after `invoice_items_total` became a `Money` object

File: app/models/registration.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (dunkOnIT) [reply app/models/registration.rb]: Changed

File: app/models/registration.rb [hunk 5]
```diff
@@ -176,6 +177,48 @@ def to_be_paid_through_wca?
...
+    )
+  end
+
+  def add_donation(iso_amount)
+    invoice_items.create(
+      amount_lowest_denomination: iso_amount,
+      currency_code: competition.currency_code,
+      display_name: I18n.t('registrations.payment_form.labels.donation'),
+    )
+  end
```
[the remark this answers is not in the document]
Comment (dunkOnIT) [reply app/models/registration.rb]: I was worried it would be complicated to try and combine them, but turns out it's not so bad.

These could be further combined into an `add_invoice_item` method - but at that point we may as well just create the invoice item directly wherever it's being called in the code. I think I like the convenience of just calling add_competition_entry or add_donation(amount) over all the variables that we'd need to pass in an invoice_item.create() call

File: app/models/registration_payment.rb [hunk 6]
```diff
@@ -32,4 +33,8 @@ def payment_status
   private def auto_close_hook
     registration.consider_auto_close
   end
+
+  private def attempt_invoice_item_status_update
+    registration.invoice_items.each { it.update(status: 'paid') } if amount_lowest_denomination == registration.invoice_items_total.cents
```
[the remark this answers is not in the document]
Comment (dunkOnIT) [reply app/models/registration_payment.rb:38]: You've just reminded me why I had the prod flags in place :facepalm: 

The intention of this PR is just a "happy-path" implementation of invoice items - I have a list of edge cases (like refunds) that I still want to introduce tests and support for, but wanted to have this basic support already in the codebase to build on top of, to keep the scope of the PR manageable

File: lib/database_dumper.rb [hunk 7]
```diff
@@ -440,6 +440,7 @@ def self.actions_to_column_sanitizers(columns_by_action)
         ),
       ),
     }.freeze,
+    "invoice_items" => :skip_all_rows,
```
[the remark this answers is not in the document]
Comment (dunkOnIT) [reply lib/database_dumper.rb:442]: We don't have the ability to conditionally fake values, do we? I'd want to give a fake amount value, but only for donations - so that we aren't publicly exposing amounts competitors may have donated to competitions. But besides that, I agree that all this info can be public.

File: spec/factories/invoice_items.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :invoice_item do
+    registration factory: %i[registration]
+    amount_lowest_denomination { 0 }
+    currency_code { registration.competition.currency_code }
+    status { 0 }
```
[the remark this answers is not in the document]
Comment (dunkOnIT) [reply spec/factories/invoice_items.rb:8]: Done

File: spec/factories/registrations.rb [hunk 9]
```diff
@@ -20,6 +20,10 @@
 
     competing_status { Registrations::Helper::STATUS_PENDING }
 
+    trait :invoice_item do
+      after(:create) { it.add_competition_entry }
```
[the remark this answers is not in the document]
Comment (dunkOnIT) [reply spec/factories/registrations.rb]: What do you mean "by default"? The intention of this trait is to add an invoice item to a registration in cases where it is desired

File: lib/database_dumper.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (gregorbg) [reply lib/database_dumper.rb]: Hm, that's an interesting thought... But I just realized that we actually do skip/clear out `registration_payments` anyways, so skipping this should be fine too. Consider my previous comment retracted :speak_no_evil:
